### PR TITLE
Refactor song select footer buttons

### LINKED ISSUE #3143: Selected mods should be more obvious on songselect screen
Not exactly obvious, I played a few maps with NoFail on and didn't realize it was still on.

image

stable has some text near the mod select button.

### LINKED ISSUE #4174: Replace with "rewind" the "random" button text while holding shift
**Describe the feature:** I was considering the idea of making the text of the “random” button more dynamic: because of its very cool rewind function using the `shift` key, I suggest to change its text to “rewind” (or something like that) when holding `shift`, and turn it back to “random” when `shift` is released. This can be useful for new players or, in general, for anyone that didn't noticed this function (and it can be visually nice) because so far it has always been very hidden as a function, so only a few people use it.
Also, if you're considering adding a sort of "hint box" (like on stable) on the main menu, maybe you can add this thing to the list.

### PR DESCRIPTION
**Pull request by LeNitrous**
This will allow certain buttons (i.e. mods and random) to have special properties and behaviors by subclassing `FooterButton`.

Closes https://github.com/ppy/osu/issues/3143, closes https://github.com/ppy/osu/issues/4174

- [x] Mods
image
based on design in https://github.com/ppy/osu/issues/3143#issuecomment-409103624
- [x] Random
based on jorolf's PR in https://github.com/ppy/osu/pull/4177

### CODE REVIEW

**Comment by LeNitrous** [comment]
Still not satisfied `.AddButton()` existing (probably made it worse by bloating it) but I'm not sure how I would approach on the refactor yet.

**Comment by LeNitrous** [comment]
Is there also planned behavior for the random button? Currently jorolf's implementation only allows keyboard to rewind.

File: osu.Game/Screens/Select/Footer.cs [hunk 1]
```diff
@@ -41,21 +39,16 @@ public class Footer : Container
         /// <para>Higher depth to be put on the left, and lower to be put on the right.</para>
         /// <para>Notice this is different to <see cref="Options.BeatmapOptionsOverlay"/>!</para>
         /// </param>
-        public void AddButton(string text, Color4 colour, Action action, Key? hotkey = null, float depth = 0)
+        public void AddButton(FooterButton button, string text, Color4 colour, Action action, Key? hotkey = null, float depth = 0)
```
Comment (jorolf) [inline osu.Game/Screens/Select/Footer.cs]: If the caller has to create the ``FooterButton`` themselves then they should probably also set the other parameters of this method in the constructor...

Comment (LeNitrous) [reply osu.Game/Screens/Select/Footer.cs]: I just left `.AddButton()` mostly as is since the buttons share the same properties.

**Comment by peppy** [comment]
Did you consider using `ModDisplay` inside the button? Might work better as it is the standardised class we used for displaying mods (and comes with expansion and tooltip logic).

**Comment by LeNitrous** [comment]
I did but the contracting and expanding animations might cause an issue for being readable (for preference I'd like all the mods to be readable). If it does I might consider adding a property to `ModDisplay` to disable it.

**Comment by LeNitrous** [comment]
Still unsure whether the animations `ModDisplay` provides is okay. I'll leave a preview here what it looks like if that was used.

https://streamable.com/kfzve

File: osu.Game/Screens/Select/Footer.cs (hunk 1 above)
Comment (peppy) [reply osu.Game/Screens/Select/Footer.cs]: Have applied this review myself.

**Review by peppy — DISMISSED** [review]

**Review by smoogipoo — APPROVED** [review]

**Pull request merged**